# SOA1039

## Don't register abstract ScheduledJob

This analyzer will generate an error if the attribute
`[ScheduledPlugIn]` (or deriving attributes) is found on an
abstract scheduled job class.

Decorating an abstract job with this attribute will
cause Optimizely to throw an exception during startup.

## Example

This will trigger SOA1039:
```C#
using EPiServer.PlugIn;
using EPiServer.Scheduler;
				
namespace SOA1039
{
	[ScheduledPlugIn(DisplayName = "Test Job", GUID = "01234567-89ab-cdef-0123-456789abcdef")]
	public abstract class MyScheduledJob : ScheduledJobBase // <-- SOA1039
	{
		// ...
	}
}
```
The fix:

Either remove the `abstract` keyword or `ScheduledPlugIn`Attribute from `MyScheduledJob`,
depending on what behavior you want.

## Related analyzers

This analyzer is a part of a group of analyzers surrounding abstract
classes with Optimizely registration attributes.

- [SOA1037 - Don't register abstract PropertyDefinitionTypePlugIn](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1037.md)
- [SOA1038 - Don't register abstract ContentType](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1038.md)
- [SOA1039 - Don't register abstract ScheduledJob](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1039.md)
- [SOA1040 - Don't register abstract Initialization module](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1040.md)

This analyzer is also part of a group of analyzers focusing
on Scheduled jobs:

- [SOA1006 - Inherit from ScheduledJobBase](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1006.md)
- [SOA1007 - Decorate with ScheduledPluginAttribute](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1007.md)
- [SOA1008 - ScheduledPluginAttribute has no GUID](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1008.md)
- [SOA1011 - ScheduledPluginAttribute has an invalid GUID](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1011.md)
- [SOA1012 - Multiple scheduled jobs must not share GUID](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1012.md)
- [SOA1039 - Don't register abstract ScheduledJob](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1039.md)
- [SOA1041 - ScheduledJob has no name](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1041.md)
 
## Read more

- [SOA Documentation - Observation on Attributes and Abstracts](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Other/Observation-on-attributes-and-abstracts.md)