# SOA1019

## Initialize property inside SetDefaultValues

This analyzer will generate a warning if a content property is initialized
with a default value in the same way properties normally are
assigned a default value.

The correct way to assign a default value to new content items is by overriding
the method `SetDefaultValues`. Using property initialization will not have any effect,
values assigned this way will not show up in the CMS, be rendered on the site or
persisted to the database. Properties with an initializer will not have a value
until assigned one by an editor or programmatically to an instance of the content item
after it's creation.

## Example

This will trigger SOA1019:
```C#
using EPiServer.Core;
using EPiServer.DataAnnotations;

namespace SOA1019
{
    [ContentType(GroupName = "Content", GUID = "7c74f40d-cac8-4584-b5b4-09fc3e55e2b2")]
    public class MyPage : PageData
    {
        public virtual string Heading { get; set; } = "Default heading"; // <-- SOA1019
    }
}
```

The fix (codefix available):
```C#
using EPiServer.Core;
using EPiServer.DataAnnotations;
using EPiServer.DataAbstraction;

namespace SOA1019
{
    [ContentType(GroupName = "Content", GUID = "7c74f40d-cac8-4584-b5b4-09fc3e55e2b2")]
    public class MyPage : PageData
    {
        public virtual string Heading { get; set; }

        public override void SetDefaultValues(ContentType contentType)
        {
            base.SetDefaultValues(contentType);
            Heading = "Default heading";
        }
    }
}
```

## Read more

- See my [SetDefaultValues experiment project](https://github.com/Stekeblad/SetDefaultValues-experiment)
that shows the effect of both the right and wrong way side by side.