# SOA1013

## "Only one of SelectOne, SelectMany, AutoSuggestSelection (or deriving attributes) should be used on a property

A content property can only have one selection factory and property editor
associated with it at once. It does also not make sense to combine for example
the attributes SelectOne and SelectMany, is an editing user allowed to pick
exactly one option or should they be allowed to pick anything between
zero and all options?

Ignoring this warning will result in Optimizely picking one of the attributes
and ignoring the others.

## Example

wrong

```CSharp
[Display(Name = "Select something", Order = 100, GroupName = "Content")]
[SelectOne(SelectionFactoryType = typeof(MySelectionFactory))]
[SelectMany(SelectionFactoryType = typeof(MySelectionFactory))]
[AutoSuggestSelection(typeof(MySelectionFactory))]
public virtual string SelectSomething { get; set; }
```

correct

```CSharp
[Display(Name = "Select something", Order = 100, GroupName = "Content")]
[SelectOne(SelectionFactoryType = typeof(MySelectionFactory))]
public virtual string SelectSomething { get; set; }
```

## Read more
- [SOA1015 - Attribute is missing the SelectionFactoryType parameter](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1015.md)
- [Developer documentation - Single or multiple list options](https://docs.developers.optimizely.com/content-cloud/v12.0.0-content-cloud/docs/single-or-multiple-list-options)
- [Developer documentation - Built-in auto-suggestion editor](https://docs.developers.optimizely.com/content-cloud/v12.0.0-content-cloud/docs/built-in-auto-suggestion-editor)
- [Developer documentation - Built-in property types](https://docs.developers.optimizely.com/content-cloud/v12.0.0-content-cloud/docs/built-in-property-types#select-predefined-values)