# SOA1014

## Type may not work with the attributes SelectOne, SelectMany or AutoSuggestSelection (or deriving attributes)

Not all types that can be used as properties on content can be used
together with the selection factory attributes. Types that works are
`enum` types, `string` and `int`. Limitations apply for int and enums.
The consequence of using any other type or not considering the
limitations described below can be that the selections will
not be persisted or Optimizely failing to save the entire content item.

### Limitations with int

int should only be used if editors are limited to only choosing one
value and all values from the `SelectItem`s in the selection factory
are integers. Use string for all other values or if multiple
options are allowed to be selected.

### Limitations with enums

Using enums works well with `SelectOneAttribute` assuming all selectable
values can be represented by the enum type. A bit of extra
work is required for enums to be used with `SelectManyAttribute`. See for example
[this blog post by Grzegorz Wiecheć](https://gregwiechec.com/2017/07/how-to-use-selectmany-to-work-with-flags-enum/).

## Example

wrong

```CSharp
// ButtonBlock is not an allowed type when using SelectionFactories
[Display(Name = "Select something", Order = 100, GroupName = "Content")]
[SelectOne(SelectionFactoryType = typeof(MySelectionFactory))]
public virtual ButtonBlock SelectSomething { get; set; }
```

correct

```CSharp
[Display(Name = "Select something", Order = 100, GroupName = "Content")]
[SelectOne(SelectionFactoryType = typeof(MySelectionFactory))]
public virtual string SelectSomething { get; set; }
```

## Analyzer change log

Support for enum types was added in version 1.3.1
[View the 1.3.0 version](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/v1.3.0/doc/Analyzers/SOA1014.md)

Support for nullable types was added in version 1.5.0

## Read more
- [Developer documentation - Built-in property types](https://docs.developers.optimizely.com/content-cloud/v12.0.0-content-cloud/docs/built-in-property-types#select-predefined-values)
- [Developer documentation - Single or multiple list options](https://docs.developers.optimizely.com/content-cloud/v12.0.0-content-cloud/docs/single-or-multiple-list-options)
- [Developer documentation - Built-in auto-suggestion editor](https://docs.developers.optimizely.com/content-cloud/v12.0.0-content-cloud/docs/built-in-auto-suggestion-editor)
- [Grzegorz Wiecheć - How to use SelectMany to work with flags enum](https://gregwiechec.com/2017/07/how-to-use-selectmany-to-work-with-flags-enum/)