# SOA1041

## ScheduledJob has no name

This analyzer will generate a warning if the attribute
`[ScheduledPlugin]` has no `DisplayName` or
[LanguagePath](https://world.optimizely.com/csclasslibraries/cms/EPiServer.PlugIn.PlugInAttribute?version=12#EPiServer_PlugIn_PlugInAttribute_LanguagePath)
parameter. This analyzer will also trigger if `DisplayName` is null
or an empty string or if `LanguagePath` is null. If one of them
is valid it will not complain because of the other, only one of them is needed.

If no valid value is present for any of the properties, then, a new job
can fail to register in Optimizely and an already existing job can in worst
case disappear and stop running. Read more further down.

## Example

This will trigger SOA1041:
```C#
using EPiServer.PlugIn;
using EPiServer.Scheduler;
				
namespace SOA1041
{
	// SOA1041: The attribute have no DisplayName or LanguagePath parameter
	[ScheduledPlugIn(GUID = "01234567-89ab-cdef-0123-456789abcdef")]
	public class MyScheduledJob : ScheduledJobBase
	{
		// ...
	}
}
```

The fix:

Add either a non-empty `DisplayName` or `LanguagePath` parameter to the `ScheduledPlugIn`Attribute

## The importance of a name

Scheduled jobs need a name to show in the admin interface and Optimizely offers
two ways to assign one. The first one is simple, set the name you want to show
in the `DisplayName` property of `ScheduledPlugInAttribute` and that is the
name that will show up. The second way is to set the `LanguagePath`
property on the attribute instead and Optimizely will use that string to
retrieve a translatable name from the `LocalizationService`.

If you create a new scheduled job and do not set either of these two properties
on the attribute you will find that your job will not be registered, it does
not show up in the admin interface and it does not run even if you set it to be
enabled by default and specify an interval on `ScheduledPlugInAttribute`.

If you have a job that has been registered but no longer have a `DisplayName` or
`LanguagePath` then it will not properly update if some of it's metadata changes.

Two examples of what can happen are the following (not exhaustive):

1. The job will not react when you add or remove the
`IsStoppable` capability. Much of the metadata about a scheduled job can
be set on `ScheduledPlugInAttribute` but Stoppable is interesting because it
does not have anything to do with it. To make a job stoppable you set the
`IsStoppable` property from the `ScheduledJobBase` base class to true in the
constructor and override the `Stop()` method.

1. Optimizely will not pick up that the job has been renamed or moved
to another namespace. It will behave as you have deleted the job and created
a new one, and because the job does not have a name the new job can't be
registered and therefore it will be impossible to run it.
[Assigning a GUID](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1008.md)
to the job does not help in this case.

### Is my site affected?

If you have access to the logs, look for a message like the below during startup:

```
EPiServer.Scheduler.Internal.DefaultScheduledJobScanner[0]
Failed to register scheduled job 'MySite.Business.ScheduledJobs.DoSomethingJob'
EPiServer.DataAbstraction.DataAbstractionException: Cannot save a job without name
```

If you have access to the code you can inspect all your scheduled jobs to see
if any of them is missing those very important `DisplayName` or `LanguagePath`
parameters on the `ScheduledPlugIn` Attribute.
You could also install the nuget package Stekeblad.Optimizely.Analyzers version 1.9.0 or later
and look for warnings with ID SOA1041, it will also help you find many other
issues specific to Optimizely sites.

## Related analyzers

This analyzer is a part of a group of analyzers focusing
on Scheduled jobs:

- [SOA1006 - Inherit from ScheduledJobBase](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1006.md)
- [SOA1007 - Decorate with ScheduledPluginAttribute](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1007.md)
- [SOA1008 - ScheduledPluginAttribute has no GUID](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1008.md)
- [SOA1011 - ScheduledPluginAttribute has an invalid GUID](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1011.md)
- [SOA1012 - Multiple scheduled jobs must not share GUID](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1012.md)
- [SOA1039 - Don't register abstract ScheduledJob](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1039.md)
- [SOA1041 - ScheduledJob has no name](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1041.md)