# SOA1001

## Use ContentTypeAttribute

Content types must be decorated with 
[`ContentTypeAttribute`](https://docs.developers.optimizely.com/content-management-system/docs/attributes)
(or an attribute inheriting from it)
in order to be discovered.

## Multiple ways to fix

There are three different ways to fix this diagnostic and all
with different results.

1. Add the `ContentTypeAttribute` attribute, the content type will show up
for editors in the CMS.
2. Add the `ContentTypeAttribute` attribute and set the `AvailableInEditMode`
property to `false`. The class will become usable as a content type but
editors will not be able to create new instances of it.
3. Add the `abstract` keyword to the class declaration, signaling this
class is a base class other classes build on and not to be directly instantiated.

## Example

Take the below code section as an example. It contains three locations diagnosing with SOA1001
and in this example they will be fixed with one each of the ways detailed above.

```C#
// namespace and usings omitted

public class TopImageBlock : BlockData
{
	[UIHint(UIHint.Image)]
	public virtual ContentReference TopImage { get; set; }
	public virtual string Description { get; set; }
}

public class SiteBasePage : PageData
{
	public virtual string Heading { get; set; }
}

public class StandardPage : SitePageData
{
	public virtual TopImageBlock TopImage { get; set; }
	public virtual XhtmlString Body { get; set; }
}
```

Let's begin with `SitePageData`, it is intended to be a base type for other pages
to inherit from but not to be instantiated directly. To make this intention clearer,
I'll change `SitePageData` to be abstract.

Next `StandardPage`. On the site in this example it will be one of the most common
content types on the site. Here we must add the `ContentType` attribute on the
class so it becomes available to use by editors.

Lastly `TopImageBlock`. This block is not intended to be created by editors but
it is also not a base type that can be declared as abstract. This block type
is used as a property on StandardPage, requiring it to be registered as a
content type. To prevent the type to be used as more than a property on
other types I'll add the `ContentType` attribute and set `AvailableInEditMode`
property to `false`.

Now the example snippet above looks like this:

```C#
// namespace and usings omitted

[ContentType(AvailableInEditMode = false, GUID = "ABCDABCD-1111-ABCD-ABCD-ABCDABCDABCD")]
public class TopImageBlock : BlockData
{
	[UIHint(UIHint.Image)]
	public virtual ContentReference TopImage { get; set; }
	public virtual string Description { get; set; }
}

public abstract class SiteBasePage : PageData
{
	public virtual string Heading { get; set; }
}

[ContentType(GUID = "ABCDABCD-ABCD-ABCD-3333-ABCDABCDABCD")]
public class StandardPage : SitePageData
{
	public virtual TopImageBlock TopImage { get; set; }
	public virtual XhtmlString Body { get; set; }
}
```

## Tips

The `GroupName` parameter can be used to group releated content types

Define the name and description of the content type using the
[language files xml](https://docs.developers.optimizely.com/content-cloud/v12.0.0-content-cloud/docs/localizing-the-user-interface)
to support localizing them to different languages instead of using the properties on the attribute.

You have many more options then just `AvailableInEditMode` when it comes to controlling the creation of new instances of
a content type. You can use the [`AccessAttribute`](https://docs.developers.optimizely.com/content-management-system/docs/attributes#access)
to limit creation to certain roles or uses. You can use
[`AvailableContentTypesAttribute`](https://docs.developers.optimizely.com/content-management-system/docs/attributes#availablecontenttypesattribute)
to control when new instances of the content type can be created based on the type of it's parent.
To limit access to existing instances, use
[set access rights](https://support.optimizely.com/hc/en-us/articles/4413200626829-Access-rights)
in the admin interface.

You can set `AvailableInEditMode` or add/remove the attributes mentioned
in the previous paragraph at any time. Use this to for example
hide special types like the start page once the site is set up
and running to make it easier for editors find the
type they want to use in the list of available types.

## Related analyzers

This analyzer is a part of a group of analyzers surrounding
the `ContentType` attribute.

- [SOA1001 - Use ContentTypeAttribute](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1001.md)
- [SOA1002 - Add GUID parameter to ContentTypeAttribute](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1002.md)
- [SOA1009 - Multiple content types must not share GUID](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1009.md)
- [SOA1010 - ContentTypeAttribute has an invalid GUID](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1010.md)
- [SOA1038 - Don't register abstract ContentType](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1038.md)

## Read more
- [SOA1002 - Add GUID parameter to ContentTypeAttribute](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1002.md)
- [Developer documentation - ContentType attribute](https://docs.developers.optimizely.com/content-management-system/docs/attributes)
- [Class library - ContentTypeAttribute](https://world.optimizely.com/CsClassLibraries/cms/EPiServer.DataAnnotations.ContentTypeAttribute?version=12)