# SOA1015

## Attribute is missing the SelectionFactoryType parameter

This is reported for usages of the SelectOne or SelectMany attributes
that does not set a value for the attribute parameter
`SelectionFactoryType`. This parameter is required for Optimizely to know
what options should be available to select for the property.

For example, the attribute SelectOne itself tells Optimizely that editors
must "pick exactly one of the following options" as the value for
the decorated property. But what is "The following options"?
That question is answered by the atribute parameter SelectionFactoryType.
SelectionFactoryType must be a type that implements `ISelectionFactory`.

Unlike some of the other analyzers related to selection factories does this
one not report on custom attributes inheriting from SelectOne or SelectMany
as custom attributes may override the SelectionFactoryType and define a
fixed or default value,
[see this example in the Developer Documentation](https://docs.developers.optimizely.com/content-cloud/v12.0.0-content-cloud/docs/single-or-multiple-list-options#creating-your-own-attributes).

## Example

### Mininal example

wrong

```CSharp
[SelectOne]
public virtual string SelectSomething { get; set; }
```

correct

```CSharp
[SelectOne(SelectionFactoryType = typeof(MySelectionFactory))]
public virtual string SelectSomething { get; set; }
```

### Complete example

```CSharp
using EPiServer.Core;
using EPiServer.DataAnnotations;
using EPiServer.Shell.ObjectEditing;
using System.ComponentModel.DataAnnotations;
using System.Collections.Generic;

namespace SOA1015
{
    [ContentType(GUID = "9DDC8A41-5C8C-4BE0-8E73-520FF3DE8267")]
    public class ExamplePage : PageData
    {
        [Display(Name = "Select a prize", Order = 100, GroupName = "Content")]
        [SelectOne(SelectionFactoryType = typeof(MedalSelectionFactory))]
        public virtual string Medal { get; set; }
    }

    public class MedalSelectionFactory : ISelectionFactory
    {
        public IEnumerable<ISelectItem> GetSelections(ExtendedMetadata metadata)
        {
            return new List<ISelectItem>
            {
                new SelectItem { Text = "Gold", Value = "first-place" },
                new SelectItem { Text = "Silver", Value = "second-place" },
                new SelectItem { Text = "Bronze", Value = "third-place" }
            };
        }
    }
}
```

## Read more
- [Developer documentation - Single or multiple list options](https://docs.developers.optimizely.com/content-cloud/v12.0.0-content-cloud/docs/single-or-multiple-list-options)
- [Developer documentation - Built-in auto-suggestion editor](https://docs.developers.optimizely.com/content-cloud/v12.0.0-content-cloud/docs/built-in-auto-suggestion-editor)
- [Developer documentation - Built-in property types](https://docs.developers.optimizely.com/content-cloud/v12.0.0-content-cloud/docs/built-in-property-types#select-predefined-values)