# SOA1037

## Don't register abstract PropertyDefinitionTypePlugIn

This analyzer will generate an error if the attribute
`[PropertyDefinitionTypePlugIn]` is found on an abstract class
directly or indirectly deriving from `EPiServer.Core.PropertyData`.

Decorating an abstract property definition type with this attribute will
cause Optimizely to throw an exception during startup.

## Example

This will trigger SOA1037:
```C#
using EPiServer.Core;
using EPiServer.PlugIn;

namespace SOA1037
{
    [PropertyDefinitionTypePlugIn]
    public abstract class Myproperty : PropertyLongString // <-- SOA1037
    {
        // ...
    }
}
```

The fix:

Either remove the `abstract` keyword or `PropertyDefinitionTypePlugIn`Attribute from `Myproperty`,
depending on what behavior you want.

## Related analyzers

This analyzer is part of a group of analyzers for the `PropertyDefinitionTypePlugInAttribute` attribute

- [SOA 1030 - Missing PropertyDefinitionTypePlugInAttribute on custom property definition](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1030.md)
- [SOA 1031 - Missing or incompatible base class for property definition](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1031.md)
- [SOA 1032 - Add GUID to PropertyDefinitionTypePlugInAttribute](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1032.md)
- [SOA 1033 - PropertyDefinitionTypePlugInAttribute has an invalid GUID](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1033.md)
- [SOA 1034 - Multiple PropertyDefinitionTypePlugInAttributes must not share GUID](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1034.md)
- [SOA 1037 - Don't register abstract PropertyDefinitionTypePlugIn](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1037.md)

This analyzer is also part of a group of analyzers surrounding abstract
classes with Optimizely registration attributes.

- [SOA1037 - Don't register abstract PropertyDefinitionTypePlugIn](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1037.md)
- [SOA1038 - Don't register abstract ContentType](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1038.md)
- [SOA1039 - Don't register abstract ScheduledJob](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1039.md)
- [SOA1040 - Don't register abstract Initialization module](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1040.md)

## Read more

- [SOA Documentation - Observation on Attributes and Abstracts](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Other/Observation-on-attributes-and-abstracts.md)