# SOA1002

## Add GUID parameter to ContentTypeAttribute

By assigning a GUID to your content type you enable Optimizely to track it
even if it at some point is renamed or moved to a different namespace.

Renaming or moving a content type without a GUID bears a high risk of
Optimizely mistaking it for a new content type and that can lead to
data loss.

## Example

```CSharp
[ContentType] // SOA1002: Missing GUID property, continue reading documentation before adding!
public class MyPage : PageData
{
}
```

## Important

Unless the content type is brand new, DO NOT JUST ADD A RANDOM GUID!

If the content type exists in the database it has already been assigned a random
GUID by Optimizely. Locate the content type in the administrator interface
and click on the Settings button to view its assigned GUID.
Add this GUID to the attribute.

Assume an automatically generated GUID to be generated differently in
all environments. If this content type exists in production, look up the
GUID there!

Adding a new random GUID to a content type that has already been assigned
a GUID will not change the GUID in the database. It will only be misleading
as it will not enable safe moving or renaming of the content type,
that will only work if the GUID parameter on the ContentTypeAttribute
exactly match the GUID in the database.

## Note

This analyzer also trigger on attributes inheriting from ContentTypeAttribute
that do not have a GUID parameter.

## Related analyzers

This analyzer is a part of a group of analyzers surrounding
the `ContentType` attribute.

- [SOA1001 - Use ContentTypeAttribute](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1001.md)
- [SOA1002 - Add GUID parameter to ContentTypeAttribute](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1002.md)
- [SOA1009 - Multiple content types must not share GUID](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1009.md)
- [SOA1010 - ContentTypeAttribute has an invalid GUID](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1010.md)
- [SOA1038 - Don't register abstract ContentType](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1038.md)

## Read more
- [Developer documentation - Content Type](https://docs.developers.optimizely.com/content-cloud/v12.0.0-content-cloud/docs/content-types)
- [Developer documentation - Refactoring content type classes](https://docs.developers.optimizely.com/content-cloud/v12.0.0-content-cloud/docs/refactoring-content-type-classes)
- [Class library - ContentTypeAttribute](https://world.optimizely.com/CsClassLibraries/cms/EPiServer.DataAnnotations.ContentTypeAttribute?version=12)