# SOA1038

## Don't register abstract ContentType

This analyzer will generate a warning if the attribute
`[ContentType]` (or deriving attributes) is found on an abstract class
directly or indirectly deriving from `EPiServer.Core.ContentData`.

Decorating an abstract ContentType with this attribute will
not cause Optimizely to throw an exception during startup, unlike
several of the other class-attribute pairs covered by Stekeblad.Optimizely.Analyzers.

However, decorating an abstract content type with this attribute sends
confusing signals to other developers reading the code. The keyword `abstract`
sends the signal it's a base type and it cannot be instantiated while the attribute
indicates you intend to create instances of it and want Optimizely to register the type.

## Example

This will trigger SOA1038:
```C#
using EPiServer.Core;
using EPiServer.DataAnnotations;

namespace SOA1038
{
    [ContentType(Name = "Example Page", GUID = "12341234-1234-1234-1234-123412341234")]
    public abstract class ExamplePage : PageData // <-- SOA1038
    {
        // ...
    }
}
```
The fix:

Either remove the `abstract` keyword or `ContentType` attribute from `ExamplePage`,
depending on what behavior you want.

## Related analyzers

This analyzer is a part of a group of analyzers surrounding
the `ContentType` attribute.

- [SOA1001 - Use ContentTypeAttribute](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1001.md)
- [SOA1002 - Add GUID parameter to ContentTypeAttribute](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1002.md)
- [SOA1009 - Multiple content types must not share GUID](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1009.md)
- [SOA1010 - ContentTypeAttribute has an invalid GUID](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1010.md)
- [SOA1038 - Don't register abstract ContentType](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1038.md)

This analyzer is also part of a group of analyzers surrounding abstract
classes with Optimizely registration attributes.

- [SOA1037 - Don't register abstract PropertyDefinitionTypePlugIn](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1037.md)
- [SOA1038 - Don't register abstract ContentType](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1038.md)
- [SOA1039 - Don't register abstract ScheduledJob](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1039.md)
- [SOA1040 - Don't register abstract Initialization module](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1040.md)

## Read more

- [SOA Documentation - Observation on Attributes and Abstracts](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Other/Observation%20on%20attributes%20and%20abstracts.md)