# SOA1030

## Missing PropertyDefinitionTypePlugInAttribute on custom property definition

Custom property types (classes directly or indirectly inheriting from
`EPiServer.Core.PropertyData`) must be decorated with `EPiServer.PlugIn.PropertyDefinitionTypePlugInAttribute`
for Optimizely to register them as property types.

If the reported class is only intended as a base class for other property definitions -
consider marking the class as abstract, that will also remove this warning.

## Example

```C#
// SOA1030: Missing PropertyDefinitionTypePlugInAttribute
public class MyPropertyDefinition : EPiServer.Core.PropertyData
{
}

// OK: Marked as abstract and ignored
public abstract class MyBasePropertyDefinition : EPiServer.Core.PropertyLongString
{
}

// OK: Attribute present
[EPiServer.PlugIn.PropertyDefinitionTypePlugIn(GUID = "87ca4b10-dbc5-4e0a-ae23-c79a113ef00a")]
public class MyPropertyDefinition : EPiServer.Core.PropertyList<SomeType>
{
}
```


## Related analyzers

This analyzer is part of a group of analyzers for the `PropertyDefinitionTypePlugInAttribute` attribute

- [SOA 1030 - Missing PropertyDefinitionTypePlugInAttribute on custom property definition](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1030.md)
- [SOA 1031 - Missing or incompatible base class for property definition](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1031.md)
- [SOA 1032 - Add GUID to PropertyDefinitionTypePlugInAttribute](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1032.md)
- [SOA 1033 - PropertyDefinitionTypePlugInAttribute has an invalid GUID](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1033.md)
- [SOA 1034 - Multiple PropertyDefinitionTypePlugInAttributes must not share GUID](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1034.md)
- [SOA 1037 - Don't register abstract PropertyDefinitionTypePlugIn](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1037.md)

## Read more

- [Developer documentation - Generic PropertyList](https://docs.developers.optimizely.com/content-management-system/docs/generic-propertylist)
- [Developer documentation - Custom properties](https://docs.developers.optimizely.com/content-management-system/docs/custom-properties)