# SOA1016

## SelectionFactoryType does not implement ISelectionFactory or is declared abstract

This is reported for usages of the SelectOne or SelectMany attributes
where the type given as the value of the parameter `SelectionFactoryType`
does not implement `ISelectionFactory`.

Passing a type to `SelectionFactoryType` that does not implement
`ISelectionFactory` can result in that it becomes impossible to edit all
instances of the content type (and types derived from it)
with the reported property in the CMS until
the `SelectionFactoryType` parameter is updated with a valid type.

## Example

wrong

```CSharp
[SelectOne(SelectionFactoryType = typeof(MyStringExtensions))]
public virtual string SelectSomething { get; set; }
```

correct

```CSharp
// MySelectionFactory implements ISelectionFactory
[SelectOne(SelectionFactoryType = typeof(MySelectionFactory))]
public virtual string SelectSomething { get; set; }
```

## Read more
- [SOA1015 - Attribute is missing the SelectionFactoryType parameter](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1015.md)
- [Developer documentation - Single or multiple list options](https://docs.developers.optimizely.com/content-cloud/v12.0.0-content-cloud/docs/single-or-multiple-list-options)
- [Developer documentation - Built-in auto-suggestion editor](https://docs.developers.optimizely.com/content-cloud/v12.0.0-content-cloud/docs/built-in-auto-suggestion-editor)
- [Developer documentation - Built-in property types](https://docs.developers.optimizely.com/content-cloud/v12.0.0-content-cloud/docs/built-in-property-types#select-predefined-values)