# SOA1017

## SelectionFactoryType does not implement ISelectionQuery or is declared abstract

`AutoSuggestSelectionAttribute` has one required constructor parameter
that takes a `type`. This analyzer is reporting when said type does not
implement `ISelectionQuery`.

Passing a type that does not implement
`ISelectionQuery` can result in that it becomes impossible to edit all
instances of the content type (and types derived from it)
with the reported property in the CMS until
the constructor parameter is updated with a valid type.

## Example

wrong

```CSharp
[AutoSuggestSelection(typeof(MySelectionFactory))]
public virtual string SelectSomething { get; set; }
```

correct

```CSharp
// MySelectionQuery implements ISelectionQuery
[AutoSuggestSelection(typeof(MySelectionQuery))]
public virtual string SelectSomething { get; set; }
```

## Read more
- [SOA1015 - Attribute is missing the SelectionFactoryType parameter](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1015.md)
- [Developer documentation - Single or multiple list options](https://docs.developers.optimizely.com/content-cloud/v12.0.0-content-cloud/docs/single-or-multiple-list-options)
- [Developer documentation - Built-in auto-suggestion editor](https://docs.developers.optimizely.com/content-cloud/v12.0.0-content-cloud/docs/built-in-auto-suggestion-editor)
- [Developer documentation - Built-in property types](https://docs.developers.optimizely.com/content-cloud/v12.0.0-content-cloud/docs/built-in-property-types#select-predefined-values)