# SOA1031

## Missing or incompatible base class for property definition

This issue is reported if a class decorated with `EPiServer.PlugIn.PropertyDefinitionTypePlugInAttribute`
does not directly or indirectly inherit from `EPiServer.Core.PropertyData`

## Example

```C#
[EPiServer.PlugIn.PropertyDefinitionTypePlugIn(GUID = "87ca4b10-dbc5-4e0a-ae23-c79a113ef00a")]
public class MyPropertyDefinition : StartPage // SOA1031: StartPage does not inherit from PropertyData
{
}

[EPiServer.PlugIn.PropertyDefinitionTypePlugIn(GUID = "87ca4b10-dbc5-4e0a-ae23-c79a113ef00a")]
public class MyPropertyDefinition : EPiServer.Core.PropertyLongString // OK
{
}
```

## Related analyzers

This analyzer is part of a group of analyzers for the `PropertyDefinitionTypePlugInAttribute` attribute

- [SOA 1030 - Missing PropertyDefinitionTypePlugInAttribute on custom property definition](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1030.md)
- [SOA 1031 - Missing or incompatible base class for property definition](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1031.md)
- [SOA 1032 - Add GUID to PropertyDefinitionTypePlugInAttribute](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1032.md)
- [SOA 1033 - PropertyDefinitionTypePlugInAttribute has an invalid GUID](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1033.md)
- [SOA 1034 - Multiple PropertyDefinitionTypePlugInAttributes must not share GUID](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1034.md)

## Read more

- [Developer documentation - Generic PropertyList](https://docs.developers.optimizely.com/content-management-system/docs/generic-propertylist)
- [Developer documentation - Custom properties](https://docs.developers.optimizely.com/content-management-system/docs/custom-properties)