# SOA1034

## Multiple PropertyDefinitionTypePlugInAttributes must not share GUID

The GUID parameter was added in EPiServer.CMS.Core 11.14.0.
This analyzer will not be enabled if a version prior to that is found.

Note: Only GUIDs on `PropertyDefinitionTypePlugInAttribute`s are examined by
this analyzer and only usages in source can be discovered by it.

Multiple PropertyDefinitionTypePlugInAttributes can not have the same GUID, they must all be unique

## Example

```C#
// SOA1034 Multiple PropertyDefinitionTypePlugInAttributes must not share GUID

[EPiServer.PlugIn.PropertyDefinitionTypePlugIn(GUID = "87ca4b10-dbc5-4e0a-ae23-c79a113ef00a")]
public class MyPropertyDefinition1 : EPiServer.Core.PropertyLongString
{
}

[EPiServer.PlugIn.PropertyDefinitionTypePlugIn(GUID = "87ca4b10-dbc5-4e0a-ae23-c79a113ef00a")]
public class MyPropertyDefinition2 : EPiServer.Core.PropertyLongString
{
}
```

## Related analyzers

This analyzer is part of a group of analyzers for the `PropertyDefinitionTypePlugInAttribute` attribute

- [SOA 1030 - Missing PropertyDefinitionTypePlugInAttribute on custom property definition](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1030.md)
- [SOA 1031 - Missing or incompatible base class for property definition](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1031.md)
- [SOA 1032 - Add GUID to PropertyDefinitionTypePlugInAttribute](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1032.md)
- [SOA 1033 - PropertyDefinitionTypePlugInAttribute has an invalid GUID](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1033.md)
- [SOA 1034 - Multiple PropertyDefinitionTypePlugInAttributes must not share GUID](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1034.md)


## Read more

- [SOA 1032 - Add GUID to PropertyDefinitionTypePlugInAttribute](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1032.md)
- [Developer documentation - Generic PropertyList](https://docs.developers.optimizely.com/content-management-system/docs/generic-propertylist)
- [Developer documentation - Custom properties](https://docs.developers.optimizely.com/content-management-system/docs/custom-properties)