# SOA1014

## Properties decorated with one of the attributes SelectOne, SelectMany or AutoSuggestSelection (or deriving attributes) must be of the type string or int for the selection to work

Properties decorated with one of these attributes should be of type string
or int, the consequence of using any other type can be
that the selections will not be persisted or Optimizely
failing to save the entire content item.

int should only be used if editors are limited to only choosing one
value and all values from the `SelectItem`s in the selection factory
are integers. Use string for all other values or if multiple
options are allowed to be selected.

## Example

wrong

```CSharp
// ButtonBlock is not an allowed type when using SelectionFactories
[Display(Name = "Select something", Order = 100, GroupName = "Content")]
[SelectOne(SelectionFactoryType = typeof(MySelectionFactory))]
public virtual ButtonBlock SelectSomething { get; set; }
```

correct

```CSharp
[Display(Name = "Select something", Order = 100, GroupName = "Content")]
[SelectOne(SelectionFactoryType = typeof(MySelectionFactory))]
public virtual string SelectSomething { get; set; }
```

## Read more
- [Developer documentation - Built-in property types](https://docs.developers.optimizely.com/content-cloud/v12.0.0-content-cloud/docs/built-in-property-types#select-predefined-values)
- [Developer documentation - Single or multiple list options](https://docs.developers.optimizely.com/content-cloud/v12.0.0-content-cloud/docs/single-or-multiple-list-options)
- [Developer documentation - Built-in auto-suggestion editor](https://docs.developers.optimizely.com/content-cloud/v12.0.0-content-cloud/docs/built-in-auto-suggestion-editor)