# SOA1032

## Add GUID to PropertyDefinitionTypePlugInAttribute

The GUID parameter was added in EPiServer.CMS.Core 11.14.0.
This analyzer will not be enabled if a version prior to that is found.

Like `ContentTypeAttribute` and `ScheduledPluginAttribute` does
`PropertyDefinitionTypePlugInAttribute` support a GUID parameter.
Like Content types and scheduled jobs is a GUID needed to better support moving and renaming
classes. Unlike the other types is property definitions not automatically
assigned a GUID so it's not needed to go looking for a generated one you need to match.

There is a known issue when upgrading from CMS 11 to CMS 12 that can occur if
property definitions does not have assigned GUIDs. The read more section links to
blog posts by one team who had this issue. Avoid this issue by adding GUIDs and
deploying before you start upgrading.

## Example

```CSharp
[EPiServer.PlugIn.PropertyDefinitionTypePlugIn] // SOA1032: Missing GUID property
public class MyPropertyDefinition : EPiServer.Core.PropertyLongString
{
}
```

## Related analyzers

This analyzer is part of a group of analyzers for the `PropertyDefinitionTypePlugInAttribute` attribute

- [SOA 1030 - Missing PropertyDefinitionTypePlugInAttribute on custom property definition](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1030.md)
- [SOA 1031 - Missing or incompatible base class for property definition](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1031.md)
- [SOA 1032 - Add GUID to PropertyDefinitionTypePlugInAttribute](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1032.md)
- [SOA 1033 - PropertyDefinitionTypePlugInAttribute has an invalid GUID](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1033.md)
- [SOA 1034 - Multiple PropertyDefinitionTypePlugInAttributes must not share GUID](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1034.md)

## Read more

- [Developer documentation - Generic PropertyList](https://docs.developers.optimizely.com/content-management-system/docs/generic-propertylist)
- [Developer documentation - Custom properties](https://docs.developers.optimizely.com/content-management-system/docs/custom-properties)
- [Blogpost by Nick Hamlin, Perficient - Bug Fixing: Lazy loaded property value is not supported by the current property instance](https://blogs.perficient.com/2024/07/30/lazy-loaded-property-value-is-not-supported/)
- [Blogpost by Ritu Madan, Perficient - Optimizely CMS – the Importance of GUIDs and Assembly Names](https://blogs.perficient.com/2022/08/16/optimizely-cms-the-importance-of-guids-and-assembly-names/)