# SOA1008

## ScheduledPluginAttribute has no GUID

The GUID parameter was added in EPiServer.CMS.Core 10.3.
This analyzer will not be enabled if a version prior to that is installed.

> **Important!** If you change the class or namespace name the old scheduled
job will still be active but without a settings page, you should manually
delete this job from the database in table tblScheduledItem.

Quote from [Class library](https://world.optimizely.com/CsClassLibraries/cms/EPiServer.PlugIn.ScheduledPlugInAttribute?version=12)

## Example

```CSharp
[ScheduledPlugIn] // SOA1008: Missing GUID property, continue reading documentation before adding!
public class MyScheduledJob : ScheduledJobBase
{
}
```

## The GUID

Unless the scheduled job is brand new, DO NOT JUST ADD A RANDOM GUID!

If the scheduled job exists in the database it has already been assigned a
random GUID by Optimizely. If your site runs Optimizely 12 you can see the GUID
in the url when viewing a job in the administrator interface.
On older versions is the GUID not visible
in any interface and to retrieve it you have to do one of the following things:

- Write code to get it from
[IScheduledJobRepository](https://world.optimizely.com/CsClassLibraries/cms/EPiServer.DataAbstraction.IScheduledJobRepository?version=12)
- Query the database

### Use IScheduledJobRepository

```CSharp
// Prefer Depencency injection over ServiceLocator
var jobRepo = ServiceLocator.Current.GetInstance<IScheduledJobRepository>();
foreach (var job in jobRepo.List())
{
	Debug.WriteLine($"GUID={job.ID} Full type name={job.TypeName}");
}
```

### Query the database

```SQL
SELECT pkID, TypeName FROM dbo.tblScheduledItem
```

## Related analyzers

This analyzer is a part of a group of analyzers focusing
on Scheduled jobs:

- [SOA1006 - Inherit from ScheduledJobBase](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1006.md)
- [SOA1007 - Decorate with ScheduledPluginAttribute](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1007.md)
- [SOA1008 - ScheduledPluginAttribute has no GUID](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1008.md)
- [SOA1011 - ScheduledPluginAttribute has an invalid GUID](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1011.md)
- [SOA1012 - Multiple scheduled jobs must not share GUID](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1012.md)
- [SOA1039 - Don't register abstract ScheduledJob](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1039.md)
- [SOA1041 - ScheduledJob has no name](https://github.com/Stekeblad/stekeblad.optimizely.analyzers/blob/master/doc/Analyzers/SOA1041.md)

## Read more
- [Class library - ScheduledPluginAttribute](https://world.optimizely.com/CsClassLibraries/cms/EPiServer.PlugIn.ScheduledPlugInAttribute?version=12)